core/embed/trezorhal/stm32f4/i2c_bus.c

```diff
@@ -0,0 +1,951 @@
+      // 3) START condition is not requested in the next operation
+      // 4) The next operation has the same direction
+
+      while ((bus->next_op != packet->op_count) &&
+             ((op->flags & I2C_FLAG_STOP) == 0) &&
+             (((op + 1)->flags & I2C_FLAG_START) == 0) &&
+             (((op + 1)->flags & I2C_FLAG_TX) == (op->flags & I2C_FLAG_TX))) {
+        // Move to the next operation
+        op = &packet->ops[bus->next_op++];
+
+        if (op->flags & I2C_FLAG_EMBED) {
+          bus->transfer_size += MIN(op->size, sizeof(op->data));
+        } else {
+          bus->transfer_size += op->size;
+        }
+      }
+
+      // STOP condition:
+      //  1) if it is explicitly requested
+      //  2) if it is the last operation in the packet
+      bus->stop_requested = ((op->flags & I2C_FLAG_STOP) != 0) ||
+                            (bus->next_op == packet->op_count);
+
+      // Calculate address byte
+      bus->addr_byte = packet->address << 1;
+
+      // ACK, POS, ITBUFEN flags are set based on the operation
+      if (bus->transfer_size > 0) {
+        if (op->flags & I2C_FLAG_TX) {
+          cr2 |= I2C_CR2_ITBUFEN;
+        } else if (op->flags & I2C_FLAG_RX) {
+          bus->addr_byte |= 1;  // Set RW bit to 1 (READ)
+          if (bus->transfer_size == 1) {
+            cr2 |= I2C_CR2_ITBUFEN;
+          } else if (bus->transfer_size == 2) {
+            cr1 |= I2C_CR1_POS;
+          } else if (bus->transfer_size == 3) {
+            cr1 |= I2C_CR1_ACK;
+          } else if (bus->transfer_size > 3) {
+            cr2 |= I2C_CR2_ITBUFEN;
+            cr1 |= I2C_CR1_ACK;
+          }
+        }
+      }
+
+      // Enable event and error interrupts
+      cr2 |= I2C_CR2_ITEVTEN | I2C_CR2_ITERREN;
+
+      // Generate start condition
+      // (this also clears all status flags)
+      cr1 |= I2C_CR1_START;
+
+      // Each operation has its own timeout calculated
+      // based on the number of bytes to transfer and the bus speed +
+      // expected operation overhead
+      systimer_set(bus->timer,
+                   I2C_BUS_TIMEOUT(bus->transfer_size) + packet->timeout);
+
+      // Guard time between operations STOP and START condition
+      if (bus->def->guard_time > 0) {
+        // Add 5us as a safety margin since the stop_time was set before the
+        // STOP condition was issued
+        uint16_t guard_time = bus->def->guard_time + 5;
+        while (systick_us() - bus->stop_time < guard_time)
+          ;
+      }
+    }
+
+    // Clear BTF flag
+    (void)regs->DR;
+  }
+
+  regs->CR1 = cr1;
+  regs->CR2 = cr2;
+}
+
+// Timer callback handling I2C bus timeout
+static void i2c_bus_timer_callback(void* context) {
+  i2c_bus_t* bus = (i2c_bus_t*)context;
+
+  if (bus->abort_pending) {
+    // This may be caused by the bus being busy/noisy.
+    // Reset I2C Controller
+    i2c_bus_reset(bus);
+    // Start the next packet
+    i2c_bus_head_continue(bus);
```

Comment:
Yes, because of this, I changed the systick priority to `IRQ_NORMAL` when I implemented this driver. All interrupts, except for `PendSV`, `SVCCall`, and faults, are set to `IRQ_NORMAL` now and should be in future. Without this assumption, it becomes very tricky to synchronize driver IRQ handlers and timers callbacks.

Reply:
actually, you can:

- You can call `systick_delay_us()` or `systick_delay_ms()` in the SVCall handler (syscalls) since it has the lowest priority. That's perfectly fine.

- You can call these functions in IRQ handlers, but it's crucial to limit the delay to a few or at most, tens of microseconds. While the systick counter will still increment correctly even if you wait for several milliseconds (see the implementation of `systick_delay_us()` and `systick_cycles()`), doing so will block other interrupts from being handled in time.